src/main/host/host.rs

```diff
@@ -581,6 +585,101 @@ impl Host {
+        &self,
+        protocol_type: cshadow::ProtocolType,
+        src: SocketAddrV4,
+        dst: SocketAddrV4,
+    ) -> bool {
+        let port = src.port().to_be();
+        let peer_addr = u32::from(*dst.ip()).to_be();
+        let peer_port = dst.port().to_be();
+
+        if src.ip().is_unspecified() {
+            // Check that all interfaces are available.
+            !self.localhost.borrow().as_ref().unwrap().is_associated(
+                protocol_type,
+                port,
+                peer_addr,
+                peer_port,
+            ) && !self.internet.borrow().as_ref().unwrap().is_associated(
+                protocol_type,
+                port,
+                peer_addr,
+                peer_port,
+            )
+        } else {
+            // The interface is not available if it does not exist.
+            match self.with_interface_mut(*src.ip(), |iface| {
+                iface.is_associated(protocol_type, port, peer_addr, peer_port)
+            }) {
+                Some(is_associated) => !is_associated,
+                None => false,
+            }
+        }
+    }
+
+    // Returns a random port in host byte order
+    pub fn get_random_free_port(
+        &self,
+        protocol_type: cshadow::ProtocolType,
+        interface_ip: Ipv4Addr,
+        peer: SocketAddrV4,
+    ) -> Option<u16> {
+        // we need a random port that is free everywhere we need it to be.
+        // we have two modes here: first we just try grabbing a random port until we
+        // get a free one. if we cannot find one fast enough, then as a fallback we
+        // do an inefficient linear search that is guaranteed to succeed or fail.
+
+        // if choosing randomly doesn't succeed within 10 tries, then we have already
+        // allocated a lot of ports (>90% on average). then we fall back to linear search.
+        for _ in 0..10 {
+            let random_port = self
+                .random
+                .borrow_mut(&self.root)
+                .gen_range(MIN_RANDOM_PORT..=u16::MAX);
+
+            // this will check all interfaces in the case of INADDR_ANY
+            if self.is_interface_available(
+                protocol_type,
+                SocketAddrV4::new(interface_ip, random_port),
+                peer,
+            ) {
+                return Some(random_port);
+            }
+        }
+
+        // now if we tried too many times and still don't have a port, fall back
+        // to a linear search to make sure we get a free port if we have one.
+        // but start from a random port instead of the min.
+        let start = self
+            .random
+            .borrow_mut(&self.root)
+            .gen_range(MIN_RANDOM_PORT..=u16::MAX);
+        let mut port = start;
+        loop {
+            port = if port == u16::MAX {
+                MIN_RANDOM_PORT
+            } else {
+                port + 1
+            };
+            if port == start {
+                break;
+            }
```

Comment:
What about:

```rust
let ports = MIN_RANDOM_PORT..=u16::MAX;
let start_offset = self
    .random
    .borrow_mut(&self.root)
    .gen_range(0..ports.len());
for port in ports.cycle().skip(start_offset).take(ports.len()) {
    ...
}
```

Reply:
Wouldn't the `skip` be `O(n)` in `start_offset`? Or is there a clever optimization I don't know about?

(I also thought cycle required storing all previous values of the iterator, but afaict it just clones the original iterator, so doesn't have to keep the individual values around)

Reply:
Ah, it uses  [nth](https://stackoverflow.com/questions/66287676/how-does-skip-works-in-rust), which for a range iterator would presumably be constant time

Reply:
Done. Thanks, much nicer!

Reply:
That's what I was thinking as well, but looking at it again, I don't know if the `Cycle` in the middle messes it up. `Cycle::nth()` seems to call `Iterator::advance_by(n)` on the range iterator. A normal range iterator `Range` has a specialized [`advance_by()`](https://doc.rust-lang.org/src/core/iter/range.rs.html#750) implementation, but a `RangeInclusive` seems to default to the default `Iterator` [`advance_by()`](https://doc.rust-lang.org/src/core/iter/traits/iterator.rs.html#328) implementation.

So I don't know anymore :) Maybe it's better to stick with something we know is optimal like the original implementation.

Reply:
I tried to approximate on godbolt - I don't claim to fully understand the disassembly, but I don't see anything that looks like e.g. dynamically allocating somewhere to store past 

I made a contrived example with ridiculously large values. It hangs when run, even in release builds, so it definitely seems like non-constant time after all. I'll revert that last change.

```
use std::convert::TryInto;

pub fn main() -> std::process::ExitCode {
    let ports = 0..=u64::MAX;
    let random_offset = u64::MAX/2;
    let start_offset = unsafe { std::ptr::read_volatile(&random_offset) };

    let result = ports.clone().cycle().skip(start_offset.try_into().unwrap()).take(u64::MAX.try_into().unwrap()).next().unwrap();
 
    (result as u8).into()
}
```

Reply:
Btw I just tried changing my example to use `0..u64::MAX` instead of `0..=u64::MAX`, so I think you're right that it's a problem specific to `RangeInclusive`.

I wonder if there's a fundamental reason it doesn't have an specialized `advance_by`, or whether it's just an oversight.